File: utils/cpp/src/Comm/Hub.spec.cpp
#include <catch.hpp>
#include <Comm/Hub.hpp>
#include <Comm/Bridge.mock.hpp>
#include <string>

typedef std::vector<std::string> vectString;

TEST_CASE("can emit messages", "[Hub]"){
  Comm::Mock::Bridge<std::string> bridge;
  Comm::Hub<std::string> hub(&bridge);

  SECTION("simple message"){
    hub.emit("test", vectString{"message"});
    REQUIRE( bridge.sent.size() == 1 );
    REQUIRE( bridge.sent[0] == vectString{"message", "test"} );
  }

  SECTION("complex message"){
    hub.emit("event", vectString{"data1", "data2", "data3", "data4"});
    REQUIRE( bridge.sent.size() == 1 );
    REQUIRE( bridge.sent[0] == vectString{"data1", "data2", "data3", "data4", "event"} );
  }
}

TEST_CASE("can receive messages", "[Hub]"){
  Comm::Mock::Bridge<std::string> bridge;
  Comm::Hub<std::string> hub(&bridge);

  SECTION("simple message"){
    bridge.received.push(vectString{"message", "test"});
    vectString receivedMessage;
    hub.on("test", [&receivedMessage](vectString message){
      receivedMessage = message;
    });
    hub.poll();
    REQUIRE( receivedMessage == vectString{"message"} );
  }

  SECTION("complex message"){
    bridge.received.push(vectString{"data3", "data2", "data1", "event"});
    vectString receivedMessage;
    hub.on("event", [&receivedMessage](vectString message){
      receivedMessage = message;
    });
    hub.poll();
    REQUIRE( receivedMessage == vectString{"data3", "data2", "data1"} );
  }

  SECTION("multiple handlers"){
    bridge.received.push(vectString{"data3", "data2", "data1", "event"});
    vectString receivedMessage_1, receivedMessage_2;
    hub.on("event", [&receivedMessage_1](vectString message){
      receivedMessage_1 = message;
    });
    hub.on("event", [&receivedMessage_2](vectString message){
      receivedMessage_2 = message;
    });
    hub.poll();
    REQUIRE( receivedMessage_1 == vectString{"data3", "data2", "data1"} );
    REQUIRE( receivedMessage_2 == vectString{"data3", "data2", "data1"} );
  }
}

File: utils/javascript/src/comm/hub.spec.js
const createHub = require('./hub.js')
const createMockBridge = require('../../mock').createBridge

test('hub can handle single event with one handler', done => {
  const hub = createHub(createMockBridge([
    ['event', 'item 1', 'item 2', 'item 3']
  ]))
  hub.on('event', (passedHub, data) => {
    expect(passedHub).toBe(hub)
    expect(data).toEqual(['item 1', 'item 2', 'item 3'])
    done();
  })
  hub.poll()
})

test('hub can handle single event with multiple handlers', () => {
  const hub = createHub(createMockBridge([
    ['event', 'item 1', 'item 2', 'item 3']
  ]))
  let count = 0
  hub.on('event', (passedHub, data) => count++)
  hub.on('event', (passedHub, data) => count++)
  hub.poll()
  expect(count).toBe(2)
})

test('hub can handle multiple events with multiple handlers', () => {
  const hub = createHub(createMockBridge([
    ['3', 'item 1', 'item 2', 'item 3'],
    ['2', 'item 1', 'item 2'],
    ['1', 'item 1']
  ]))
  let count = 0
  hub.on('1', (passedHub, data) => {
    expect(data).toEqual(['item 1'])
    count++
  })
  hub.on('2', (passedHub, data) => {
    expect(data).toEqual(['item 1', 'item 2'])
    count++
  })
  hub.on('3', (passedHub, data) => {
    expect(data).toEqual(['item 1', 'item 2', 'item 3'])
    count++
  })
  hub.poll()
  expect(count).toBe(3)
})

test('hub can send an event', () => {
  const bridge = createMockBridge()
  const hub = createHub(bridge)
  hub.emit('event', ['1', '2', '3', '4'])
  expect(bridge.writeData).toEqual([['event', '1', '2', '3', '4']])
})
